### PR TITLE
Add MEGASCALE multi-slice env var injection

### PR DESCRIPTION
**Pull request by ryanaoleary**
This PR introduces support for JAX Multi-Slice (Megascale) training environments and implements a stateless indexing strategy compatible with KubeRay v1.5+.

When the `MEGASCALE_NUM_SLICES` env var  is present, the webhook automatically calculates and injects the necessary coordination variables - MEGASCALE_SLICE_ID, MEGASCALE_COORDINATOR_ADDRESS, and MEGASCALE_PORT (defaulting to 8081)—required for multi-slice communication. These values are dynamically set based on the worker index and KubeRay worker group name. 

This PR does not add injection logic for `MEGASCALE_NUM_SLICES`, since this variable should be set by the user depending on the number of TPU slices required for their workload. This variable cannot be inferred reliably, because the value of `replicas` is patched by the Ray autoscaler and even for static clusters a user may choose to specify a # of slices less than the existing number of `replicas` in the worker group for their Ray workload. `MEGASCALE_NUM_SLICES` can be specified by the user in the Pod spec, and in the future will be set automatically from Ray when a multi-slice workload is specified with `num_slices` in `SlicePlacementGroup`.

This PR also updates the go dependencies to KubeRay v1.5.

This PR has been tested with:
- [X] Unit Tests
- [X] Manual Tests

### CODE REVIEW

**Comment by ryanaoleary** [comment]
cc: @andrewsykim

File: main.go [hunk 1]
```diff
@@ -733,6 +780,38 @@ func (t *TPUWebhookServer) mutatePod(admissionReview *admissionv1.AdmissionRevie
 		container := containers[i]
 		if containerRequestingTPUs(container) {
 			path := fmt.Sprintf("/spec/containers/%d/env", i)
+			envArrayExists := len(container.Env) > 0
+
+			// Multi-Slice variable injection logic.
+			if _, exists := getEnvironmentVariable("MEGASCALE_NUM_SLICES", container); exists {
```
Comment (andrewsykim) [inline main.go:786]: Will enabling multi-slice require explicitly setting the `MEGASCALE_NUM_SLICES` env var, or will we automatically detect it based on `replicas`?

**Comment by andrewsykim** [comment]
Tried to test this on a multislice TPU cluster and ran into some issues. Logs here: https://gist.github.com/andrewsykim/40a3ae43421caa7e0ffb6e25154d1b52

Here's the RayCluster:
```
apiVersion: ray.io/v1
kind: RayCluster
metadata:
  name: raycluster-tpu-v6e-multihost
spec:
  rayVersion: '2.46.0'
  headGroupSpec:
    rayStartParams: {}
    template:
      spec:
        hostNetwork: true
        dnsPolicy: ClusterFirstWithHostNet
        containers:
        - name: ray-head
          image: us-central1-docker.pkg.dev/andrewsy-gke-dev/andrewsy-gke-dev/uber-tpu-poc:v4
          imagePullPolicy: IfNotPresent
          resources:
            limits:
              cpu: "4"
              memory: 24G
            requests:
              cpu: "4"
              memory: 24G
          ports:
          - containerPort: 6379
            name: gcs
          - containerPort: 8265
            name: dashboard
          - containerPort: 10001
            name: client
          - containerPort: 8000
            name: serve
          - containerPort: 8888
            name: grpc
  workerGroupSpecs:
  - groupName: tpu-group
    replicas: 2
    minReplicas: 0
    maxReplicas: 2
    numOfHosts: 4
    rayStartParams: {}
    template:
      spec:
        hostNetwork: true
        dnsPolicy: ClusterFirstWithHostNet
        containers:
        - name: ray-worker
          image: us-central1-docker.pkg.dev/andrewsy-gke-dev/andrewsy-gke-dev/uber-tpu-poc:v4
          imagePullPolicy: IfNotPresent
          resources:
            limits:
              cpu: "24"
              google.com/tpu: "4"
              memory: 200G
            requests:
              cpu: "24"
              google.com/tpu: "4"
              memory: 200G
          env:
          - name: NODE_IP
            valueFrom:
              fieldRef:
                fieldPath: status.hostIP
          - name: VBAR_CONTROL_SERVICE_URL
            value: $(NODE_IP):8353
          - name: JAX_PLATFORMS
            value: tpu
          - name: ENABLE_PJRT_COMPATIBILITY
            value: "true"
          - name: MEGASCALE_NUM_SLICES
            value: "2"
          - name: MEGASCALE_PORT
            value: "8081"
          ports:
          - containerPort: 8081
            name: mxla
        nodeSelector:
          cloud.google.com/gke-tpu-accelerator: tpu-v6e-slice
          cloud.google.com/gke-tpu-topology: 4x4
```

And the script:
```
import os
import ray
import jax

from jax.experimental import multihost_utils

ray.init()

@ray.remote(resources={"TPU": 4})
def tpu_cores():
    multihost_utils.sync_global_devices("sync")
    print(f"TPU Worker: {os.environ.get('TPU_WORKER_ID')}")
    return f"TPU cores: {jax.device_count()}"

num_workers = int(ray.available_resources()["TPU"]) // 4
print(f"Number of TPU Workers: {num_workers}")
result = [tpu_cores.remote() for _ in range(num_workers)]
print(ray.get(result))
```

**Comment by andrewsykim** [comment]
I think there was just an issue with my script, this one seems to work using the new TPU library!

```
import ray
import jax
from ray.util.scheduling_strategies import PlacementGroupSchedulingStrategy
from ray.util.tpu import slice_placement_group

# Reserve two v6e TPU slices, each with a 4x4 topology (16 chips each).
# This topology typically has 4 VM workers, each with 4 chips.
slice_handle = slice_placement_group(topology="4x4", accelerator_version="v6e", num_slices=2)
slice_pg = slice_handle.placement_group

print("Waiting for placement group to be ready...")
ray.get(slice_pg.ready(), timeout=600) # Increased timeout for potential scaling
print("Placement group ready.")

@ray.remote(num_cpus=0, resources={"TPU": 4})
def spmd_task(world_size, rank):
    pod_name = ray.util.tpu.get_current_pod_name()
    chips_on_node = ray.util.tpu.get_num_tpu_chips_on_node()
    print(f"Worker Rank {rank}/{world_size}: Running on slice '{pod_name}' with {chips_on_node} chips.")
    print(f"TPU devices: {jax.devices()}")
    print(f"TPU cores: {jax.device_count()}")
    return rank

# Launch one task per VM in the reserved slices. The num_workers field describes the total
# number of VMs across all slices in the SlicePlacementGroup.
tasks = [
    spmd_task.options(
        scheduling_strategy=PlacementGroupSchedulingStrategy(
            placement_group=slice_pg,
        )
    ).remote(world_size=slice_handle.num_workers, rank=i)
    for i in range(slice_handle.num_workers)
]

results = ray.get(tasks)
print(f"Task results: {results}")

```

File: main.go (hunk 1 above)
Comment (andrewsykim) [reply main.go:786]: I tested and the MEGASCALE env vars only get set when setting `MEGASCALE_NUM_SLICES` is in the pod spec initially.

What's the expected behavior? Should num slices be based on `MEGASCALE_NUM_SLICES` env var or `replicas`? in the worker group? `replicas` seems cleaner but are there cases where we want to manage multiple single-slice multi-host groups in a single worker group config?

Comment (andrewsykim) [reply main.go:786]: I guess we support the latter case for autoscaling right?

Comment (ryanaoleary) [reply main.go:786]: Yeah that's the expected behavior in this PR.

I was thinking that `MEGASCALE_NUM_SLICES`, which isn't dynamically set like the other vars (i.e. `TPU_WORKER_ID`, `TPU_WORKER_HOSTNAMES` etc. which are unique values generated for a given RayCluster) should be set by either the user in their KubeRay YAML or by Ray when creating a worker in a slice placement group with `num_slices` > 1.

This PR supports the former case, for the latter I'm updating my Ray PR with the multi-slice Ray Train changes to set these env vars based on `num_slices`. I think the latter case is the ideal solution, since it's supportive of autoscaling and will integrate easily with Ray libraries. However for workloads that don't use the TPU utils, the changes in this PR enable `MEGASCALE_NUM_SLICES` to act as the on switch for multi-slice when set, and the TPU webhook will set the remaining dynamically generated env vars / Pod fields that are required for it to work end-to-end.

Comment (andrewsykim) [reply main.go:786]: okay sounds good, is there validation logic to reject a RayCluster if MEGASCALE_NUM_SLICES doesn't match `replicas`?

File: main.go [hunk 2]
```diff
@@ -688,32 +706,61 @@ func (t *TPUWebhookServer) mutatePod(admissionReview *admissionv1.AdmissionRevie
...
-	if waitTimeout(&t.wg, time.Second*1) {
-		klog.V(1).Info("MutatePod", "PodInformer AddFunc called for prior admission request")
+	// Pod indexing variables from K8s labels or assigned dynamically.
+	var replicaIndex int
+	var tpuWorkerID int
+
+	replicaIndexStr, hasReplicaLabel := pod.Labels[utils.RayWorkerReplicaIndexKey]
+	hostIndexStr, hasHostLabel := pod.Labels[utils.RayHostIndexKey]
+
+	if hasReplicaLabel && hasHostLabel {
```
Comment (andrewsykim) [inline main.go:716]: Do you think we should split this change to use new KubeRay labels into a separate PR?

File: main.go [hunk 3]
```diff
@@ -733,6 +780,38 @@ func (t *TPUWebhookServer) mutatePod(admissionReview *admissionv1.AdmissionRevie
...
+						Name:  "MEGASCALE_COORDINATOR_ADDRESS",
+						Value: coordAddress,
+					}, path, true)
+				}
+
+				// Set MEGASCALE_PORT, defaulting to 8081 since 8080 is used by Ray for metrics.
+				if val, _ := getEnvironmentVariable("MEGASCALE_PORT", container); val == "" {
+					patches, _ = addEnvVarPatch(patches, corev1.EnvVar{
+						Name:  "MEGASCALE_PORT",
+						Value: "8081",
```
Comment (andrewsykim) [inline main.go:810]: In a follow-up PR, might be good to create a `constants.go` file with all the TPU constants

**Review by andrewsykim — COMMENTED** [review]
sorry, I added these comments a couple days ago but forgot to click submit

File: main.go (hunk 2 above)
Comment (andrewsykim) [reply main.go:716]: On second thought, prbably fine to leave it

**Review by andrewsykim — COMMENTED** [review]
LGTM, few things we should try to address in follow-up PR:
* Put all TPU env var into constants, there's quite a few now and it's getting messy to inline them all
* Put some guardrails / validation in place to ensure replicas == MEGASCALE_NUM_SLICES

**Review by andrewsykim — APPROVED** [review]

File: main.go (hunk 3 above)
Comment (ryanaoleary) [reply main.go:810]: Yeah I was thinking of refactoring the webhook folder to move a lot of the logic into a `utils.go` file, I'll make a note to add that and a `constants.go` in a follow-up PR (and probably split up the test files too).

File: main.go (hunk 2 above)
Comment (ryanaoleary) [reply main.go:716]: Yeah I think it's small enough to just add it here.

File: main.go (hunk 1 above)
Comment (ryanaoleary) [reply main.go:786]: I've been thinking about whether we should add this validation, and am leaning towards that we shouldn't (or at least it should be that `MEGASCALE_NUM_SLICES` <= `replicas` or `maxReplicas` or that `MEGASCALE_NUM_SLICES`  % `maxReplicas` == 0) since this strict validation would break the autoscaling use case where `replicas` will be changed.

Another problem is that since the Ray autoscaler isn't actually aware of different slices as logical units when scaling TPU pods unless we explicitly define a unique name for the set of TPUs in the top-level labels field. Therefore for most users (unless they alter their code to only scale a specific worker group) when autoscaling with TPUs we might have multiple multi-slice configurations within the same KubeRay worker group.

This leaves two options for the user journey:
1. Users define multiple worker groups for each multi-slice configuration (even if some are identical but will be used for different jobs). Users have to define a unique multi-slice ID per worker group and then when requesting TPUs that are part of a multi-slice configuration, also specify that ID in the label selector. Alternatively, we could direct users to set `maxReplicas` to `MEGASCALE_NUM_SLICES` for each group. This would work with 0 changes to this PR for autoscaling but isn't ideal because it closely couples the application code to the KubeRay spec for anything to work.

2. We abstract this away in the webhook by allowing `MEGASCALE_NUM_SLICES` <= `replicas` or `maxReplicas`, and then when scaling we shift `MEGASCALE_SLICE_ID` to be bound from `0-MEGASCALE_NUM_SLICES-1`. The labels from KubeRay are still useful because they tell us which Pods belong to which group, but they wouldn't necessarily represent the exact index in this case if multiple multi-slice configurations scale the same worker group. For this option, there's still an issue if the user tries to scale jobs of different `MEGASCALE_NUM_SLICES`, since the Ray autoscaler isn't aware of this and if scaling the same worker group this will break our indexing logic. To solve this, users could add a `MEGASCALE_NUM_SLICES` label to the worker group and then request that number of slices as a label selector with their job.

Comment (ryanaoleary) [reply main.go:786]: I'll work on putting these issues in a doc so that it's easier to discuss them, we can probably merge this PR with the minimal changes to get static multi-slice working and then follow-up for the dynamic, multiple multi-slice configurations / autoscaling cases.
